File: mergeSort.rb
module Enumerable
	def split
		a=self.pop(self.length-self.length/2)
		return self,a
	end
end

def merge_sort array
	if array.length > 1 
		a1,a2=array.split 
		merge(merge_sort(a1),merge_sort(a2))
	elsif array.length == 1 
		return array
	end
end

def merge a1, a2
	a0=[]
	while a1.any? && a2.any?
		a0 << (a1[0] < a2[0] ? a1 : a2).shift	#compare, pop the first element with the least number, push to a0
	end
	a0.concat a2 if a1.none?
	a0.concat a1 if a2.none?
	return a0
end

p merge_sort [1]						#1
p merge_sort [1,0]						#0,1
p merge_sort [1,0,9,4]					#0,1,4,9
p merge_sort [1,0,4,9,2]				#0,1,2,4,9
p merge_sort [1,0,4,9,2,5,3,6,8,7]		#0,1,2,3,4,5,6,7,8,9

p merge [3],[2]							#2,3
p merge [3],[2,9,10]					#2,3,9,10
p merge [3,8,9],[2]						#2,3,8,9

File: Algorithms/mergeSort.js
Array.prototype.isEmpty = function(){return this.length === 0;};
  
// Merge (for merge sort)
// Take 2 array as input
function merge (firstArray, secondArray) {
  var outputArray = [];
  
  // While both input arrays are not empty
  while( !firstArray.isEmpty() && !secondArray.isEmpty() ){
  
    // Take the smallest element from both input array and compare them
    // Remove the smaller element and put it into the top of the output array
    if ( firstArray[0] <= secondArray[0] ) {
      outputArray.push( firstArray.shift() );
    }else if ( firstArray[0] > secondArray[0] ) {
      outputArray.push( secondArray.shift() );
    }
  }
  
  // After one input array is empty, put the rest of the other input into the output array
  if( firstArray.isEmpty() ){
    outputArray = outputArray.concat( secondArray );
  }else if( secondArray.isEmpty() ){
    outputArray = outputArray.concat( firstArray );
  }
  
  // Return the merged result
  return outputArray;
}


// Merge Sort
function mergeSort(array){

  // If base case is achieved, just return it
  if(array.length === 1){return array;}
  
  // Divide the array evenly into 2 parts
  var midPoint = Math.floor( array.length/2 );
  var topArray = array.splice(midPoint);
  var bottomArray = array;
  
  // Call merge sort on both parts
  topArray = mergeSort(topArray);
  bottomArray = mergeSort(bottomArray);
  
  // Merge the results of the merge sort executed on both parts
  var sortedArray = merge(topArray, bottomArray);
  
  // Return the sorted result
  return sortedArray;
}

mergeSort([6,11,2,8,99,0,-1,0.1]);
// console.log();
